### PR TITLE
send command

### PR DESCRIPTION
**Pull request by raphjaph**

### CODE REVIEW

File: src/options.rs [hunk 1]
```diff
@@ -23,6 +23,11 @@ pub(crate) struct Options {
   bitcoin_data_dir: Option<PathBuf>,
   #[clap(long, help = "Limit index to <HEIGHT_LIMIT> blocks.")]
   pub(crate) height_limit: Option<u64>,
+  #[clap(
+    long,
+    help = "Add this flag to use the send functionality of the wallet."
+  )]
+  pub(crate) reckless: bool,
```
Comment (casey) [inline src/options.rs]:
```suggestion
```

File: Cargo.toml [hunk 2]
```diff
@@ -16,7 +16,7 @@ anyhow = { version = "1.0.56", features = ["backtrace"] }
 axum = "0.5.6"
 axum-server = "0.4.0"
 bitcoin = "0.29.1"
-bitcoincore-rpc = "0.16.0"
+bitcoincore-rpc = { version = "0.16.0", git = "https://github.com/casey/rust-bitcoincore-rpc", branch = "ord" }
```
Comment (casey) [inline Cargo.toml]: Can this only be in the test server?

File: src/subcommand/wallet/send.rs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+#[derive(Debug, Parser)]
+pub(crate) struct Send {
+  ordinal: Ordinal,
+  address: String,
+}
+
+impl Send {
+  pub(crate) fn run(self, options: Options) -> Result {
+    if options.chain.network() == Network::Bitcoin {
+      bail!("Send command is not allowed on mainnet yet. Try on regtest/signet/testnet.")
```
Comment (casey) [inline src/subcommand/wallet/send.rs:15]: Also check that the bitcoin core node that we're connected to is on the right network using `getblockchaininfo` and refuse to work if it's `mainnet`.

File: src/subcommand/wallet/send.rs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    let index = Index::open(&options)?;
+    index.index()?;
+
+    let satpoint = index.find(self.ordinal.0)?.unwrap();
```
Comment (casey) [inline src/subcommand/wallet/send.rs:21]: Remove unwrap.

File: src/subcommand/wallet/send.rs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    let output = satpoint.outpoint;
+
+    let client = options.bitcoin_rpc_client()?;
+
+    let amount = client
+      .get_transaction(&output.txid, Some(true))?
+      .amount
+      .to_sat() as u64;
```
Comment (casey) [inline src/subcommand/wallet/send.rs:29]: Is this a lossy conversion?

File: src/subcommand/wallet/send.rs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    let inputs = vec![CreateRawTransactionInput {
+      txid: output.txid,
+      vout: output.vout,
+      sequence: None,
+    }];
+
+    let mut outputs = HashMap::new();
+    outputs.insert(self.address, Amount::from_sat(amount));
+
+    let tx = client.create_raw_transaction_hex(&inputs, &outputs, None, None)?;
```
Comment (casey) [inline src/subcommand/wallet/send.rs:40]: I think these can both be inlined:

```suggestion
    let tx = client.create_raw_transaction_hex(&vec![CreateRawTransactionInput {
      txid: output.txid,
      vout: output.vout,
      sequence: None,
    }], &[(self.address, Amount::from_sat(amount))].into_iter().collect(), None, None)?;
```

File: test-bitcoincore-rpc/src/lib.rs [hunk 7]
```diff
@@ -206,6 +215,33 @@ pub trait Api {
   #[rpc(name = "getblock")]
   fn getblock(&self, blockhash: BlockHash, verbosity: u64) -> Result<String, jsonrpc_core::Error>;
 
+  #[rpc(name = "createrawtransaction")]
+  fn create_raw_transaction(
+    &self,
+    utxos: Vec<CreateRawTransactionInput>,
+    outs: HashMap<String, f64>,
```
Comment (casey) [inline test-bitcoincore-rpc/src/lib.rs:222]: Floats are a bit problematic, but I remember we had difficulties with using `Amount`. I think we should try though, floats are just a bit dangerous, I always worry about rounding.

File: tests/command_builder.rs [hunk 8]
```diff
@@ -120,8 +120,10 @@ impl CommandBuilder {
     self.tempdir
   }
 
-  pub(crate) fn run(self) -> TempDir {
+  pub(crate) fn run(self) -> (TempDir, String) {
```
Comment (casey) [inline tests/command_builder.rs:123]: Maybe this should be a struct:

```
struct Output {
  tempdir: TempDir,
  stdout: String,
}
```

File: tests/wallet.rs [hunk 9]
```diff
@@ -33,3 +33,33 @@ fn list() {
...
+  rpc_server.mine_blocks(1)[0].txdata[0].txid();
+
+  let (_, stdout) = CommandBuilder::new(
+    "--chain signet wallet send 5000000000 tb1qx4gf3ya0cxfcwydpq8vr2lhrysneuj5d7lqatw",
+  )
+  .rpc_server(&rpc_server)
+  .stdout_regex(r".*")
+  .run();
+
+  let txid = rpc_server.mine_blocks(1)[0].txdata[1].txid();
```
Comment (casey) [inline tests/wallet.rs:49]:
```
rpc_server.mempool() -> Vec<Transaction>
```

**Review by casey — CHANGES_REQUESTED** [review]

File: src/subcommand/wallet/send.rs (hunk 5 above)
Comment (raphjaph) [reply src/subcommand/wallet/send.rs:29]: i64 -> u64 so I don't think so

File: test-bitcoincore-rpc/src/lib.rs (hunk 7 above)
Comment (raphjaph) [reply test-bitcoincore-rpc/src/lib.rs:222]: For this to work with `Amount` we need the `Serialize` and `Deserialize` trait on the struct in your fork of the bitcoin rust crate.

File: test-bitcoincore-rpc/src/lib.rs [hunk 10]
```diff
@@ -169,6 +179,10 @@ impl State {
 
     tx.txid()
   }
+
+  fn get_mempool(&self) -> Vec<Transaction> {
```
Comment (casey) [inline test-bitcoincore-rpc/src/lib.rs:183]:
```suggestion
  fn mempool(&self) -> &[Transaction] {
```

File: test-bitcoincore-rpc/src/lib.rs [hunk 11]
```diff
@@ -226,6 +273,47 @@ pub trait Api {
...
+      verification_progress: 0.0,
+      initial_block_download: false,
+      chain_work: Vec::new(),
+      size_on_disk: 0,
+      pruned: false,
+      prune_height: None,
+      automatic_pruning: None,
+      prune_target_size: None,
+      softforks: HashMap::new(),
+      warnings: String::from(""),
```
Comment (casey) [inline test-bitcoincore-rpc/src/lib.rs:293]:
```suggestion
      warnings: String::new(),
```

File: test-bitcoincore-rpc/src/lib.rs [hunk 12]
```diff
@@ -226,6 +273,47 @@ pub trait Api {
...
+    })
+  }
+
+  fn get_network_info(&self) -> Result<GetNetworkInfoResult, jsonrpc_core::Error> {
+    Ok(GetNetworkInfoResult {
+      version: 230000,
+      subversion: String::from(""),
```
Comment (casey) [inline test-bitcoincore-rpc/src/lib.rs:300]:
```suggestion
      subversion: String::new(),
```

File: test-bitcoincore-rpc/src/lib.rs [hunk 13]
```diff
@@ -226,6 +273,47 @@ pub trait Api {
...
+      protocol_version: 0,
+      local_services: String::from(""),
```
Comment (casey) [inline test-bitcoincore-rpc/src/lib.rs:302]:
```suggestion
      local_services: String::new(),
```

File: test-bitcoincore-rpc/src/lib.rs [hunk 14]
```diff
@@ -226,6 +273,47 @@ pub trait Api {
...
+      time_offset: 0,
+      connections: 0,
+      connections_in: None,
+      connections_out: None,
+      network_active: true,
+      networks: Vec::new(),
+      relay_fee: Amount::from_sat(0),
+      incremental_fee: Amount::from_sat(0),
+      local_addresses: Vec::new(),
+      warnings: String::from(""),
```
Comment (casey) [inline test-bitcoincore-rpc/src/lib.rs:313]:
```suggestion
      warnings: String::new(),
```

File: test-bitcoincore-rpc/src/lib.rs [hunk 15]
```diff
@@ -285,6 +373,99 @@ impl Api for Server {
...
+          previous_output: OutPoint::new(input.txid, input.vout),
+          script_sig: Script::new(),
+          sequence: Sequence(0),
+          witness: Witness::new(),
+        })
+        .collect(),
+      output: outs
+        .iter()
+        .map(|(_address, amount)| TxOut {
+          value: *amount as u64,
```
Comment (casey) [inline test-bitcoincore-rpc/src/lib.rs:401]: Is this right? If amount is an `f64` amount of BTC, and `value` is an amount in sats, this should be `(amount * COIN_VALUE) as u64`.

File: test-bitcoincore-rpc/src/lib.rs [hunk 16]
```diff
@@ -395,6 +576,10 @@ impl Handle {
     let state = self.state();
     state.blocks[&state.hashes[bi]].txdata[ti].clone()
   }
+
+  pub fn mempool(&self) -> Vec<Transaction> {
+    self.state.lock().unwrap().get_mempool()
```
Comment (casey) [inline test-bitcoincore-rpc/src/lib.rs:581]: Might also be `.into` or something else:

```suggestion
    self.state.lock().unwrap().mempool().to_vec()
```

File: tests/command_builder.rs [hunk 17]
```diff
@@ -5,6 +5,11 @@ enum ExpectedExitStatus {
   Signal(Signal),
 }
 
+pub(crate) struct Outputs {
```
Comment (casey) [inline tests/command_builder.rs:8]:
```suggestion
pub(crate) struct Output {
```

File: tests/command_builder.rs [hunk 18]
```diff
@@ -120,8 +125,13 @@ impl CommandBuilder {
     self.tempdir
   }
 
-  pub(crate) fn run(self) -> TempDir {
+  pub(crate) fn run(self) -> Outputs {
```
Comment (casey) [inline tests/command_builder.rs:128]:
```suggestion
  pub(crate) fn run(self) -> Output {
```

File: tests/command_builder.rs [hunk 19]
```diff
@@ -120,8 +125,13 @@ impl CommandBuilder {
...
     let output = self.command().output().unwrap();
-    self.check(output)
+    let stdout = String::from(str::from_utf8(&output.stdout).unwrap());
+
+    Outputs {
+      tempdir: self.check(output),
+      stdout,
+    }
```
Comment (casey) [inline tests/command_builder.rs:135]:
```suggestion
    Output {
      tempdir: self.check(output),
      stdout,
    }
```

File: tests/index.rs [hunk 20]
```diff
@@ -5,13 +5,14 @@ fn custom_index_size() {
   let rpc_server = test_bitcoincore_rpc::spawn();
   rpc_server.mine_blocks(1);
 
-  let tempdir = CommandBuilder::new("--max-index-size 1mib find 0")
+  let outputs = CommandBuilder::new("--max-index-size 1mib find 0")
```
Comment (casey) [inline tests/index.rs:8]:
```suggestion
  let output = CommandBuilder::new("--max-index-size 1mib find 0")
```

File: tests/index.rs [hunk 21]
```diff
@@ -5,13 +5,14 @@ fn custom_index_size() {
...
     .rpc_server(&rpc_server)
     .expected_stdout("4a5e1e4baab89f3a32518a88c31bc87f618f76673e2cc77ab2127b7afdeda33b:0:0\n")
     .run();
 
   assert_eq!(
-    tempdir
+    outputs
```
Comment (casey) [inline tests/index.rs:14]:
```suggestion
    output
```

File: tests/wallet.rs [hunk 22]
```diff
@@ -33,3 +33,33 @@ fn list() {
     ))
     .run();
 }
+
+#[test]
+fn send() {
+  let rpc_server = test_bitcoincore_rpc::spawn();
+  rpc_server.mine_blocks(1)[0].txdata[0].txid();
+
+  let outputs = CommandBuilder::new(
```
Comment (casey) [inline tests/wallet.rs:42]:
```suggestion
  let output = CommandBuilder::new(
```

File: tests/wallet.rs [hunk 23]
```diff
@@ -33,3 +33,33 @@ fn list() {
...
+    "--chain signet wallet send 5000000000 tb1qx4gf3ya0cxfcwydpq8vr2lhrysneuj5d7lqatw",
+  )
+  .rpc_server(&rpc_server)
+  .stdout_regex(r".*")
+  .run();
+
+  let txid = rpc_server.mempool()[0].txid();
+  assert_eq!(format!("{}\n", txid), outputs.stdout)
```
Comment (casey) [inline tests/wallet.rs:50]:
```suggestion
  assert_eq!(format!("{}\n", txid), output.stdout)
```

File: src/subcommand/wallet/send.rs [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+impl Send {
+  pub(crate) fn run(self, options: Options) -> Result {
+    if options.chain.network() == Network::Bitcoin {
+      bail!("Send command is not allowed on mainnet yet. Try on regtest/signet/testnet.")
+    }
+
+    let client = options.bitcoin_rpc_client()?;
+    if client.get_blockchain_info().unwrap().chain == "main" {
+      bail!("Send command is not allowed on mainnet yet. Try on regtest/signet/testnet.")
+    }
```
Comment (casey) [inline src/subcommand/wallet/send.rs:18]:
```suggestion
    let client = options.bitcoin_rpc_client()?;
    if options.chain.network() == Network::Bitcoin || client.get_blockchain_info().unwrap().chain == "main"{
      bail!("Send command is not allowed on mainnet yet. Try on regtest/signet/testnet.")
    }
```

File: src/subcommand/wallet/send.rs [hunk 25]
```diff
@@ -0,0 +1,54 @@
+use {super::*, bitcoin::util::amount::Amount, bitcoincore_rpc::json::CreateRawTransactionInput};
+
+#[derive(Debug, Parser)]
+pub(crate) struct Send {
+  ordinal: Ordinal,
+  address: String,
```
Comment (casey) [inline src/subcommand/wallet/send.rs:6]:
```suggestion
  address: Address,
```

File: src/subcommand/wallet/send.rs [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+    if client.get_blockchain_info().unwrap().chain == "main" {
+      bail!("Send command is not allowed on mainnet yet. Try on regtest/signet/testnet.")
+    }
+
+    let index = Index::open(&options)?;
+    index.index()?;
+
+    let output = match index.find(self.ordinal.0)? {
+      Some(satpoint) => satpoint.outpoint,
+      None => bail!(format!("Could not find {}", self.ordinal.0)),
```
Comment (casey) [inline src/subcommand/wallet/send.rs:25]:
```suggestion
      None => bail!(format!("Could not find {} in index", self.ordinal.0)),
```

File: src/subcommand/wallet/send.rs [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+    };
+
+    let amount = client
+      .get_transaction(&output.txid, Some(true))?
+      .amount
+      .to_sat()
+      .try_into()
+      .unwrap();
```
Comment (casey) [inline src/subcommand/wallet/send.rs:33]:
```suggestion
      .to_sat();
```

File: src/subcommand/wallet/send.rs [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+        sequence: None,
+      }],
+      &[(self.address, Amount::from_sat(amount))]
+        .into_iter()
+        .collect(),
+      None,
+      None,
+    )?;
+
+    let signed_tx = client.sign_raw_transaction_with_wallet(tx, None, None)?.hex;
```
Comment (casey) [inline src/subcommand/wallet/send.rs:48]:
```suggestion
    let signed_tx = client.sign_raw_transaction_with_wallet(client.create_raw_transaction_hex(
      &[CreateRawTransactionInput {
        txid: output.txid,
        vout: output.vout,
        sequence: None,
      }],
      &[(self.address, Amount::from_sat(amount))]
        .into_iter()
        .collect(),
      None,
      None,
    )?, None, None)?.hex;
```

**Review by casey — APPROVED** [review]
We have a send command!

File: test-bitcoincore-rpc/src/lib.rs (hunk 10 above)
Comment (Muhammad4599) [reply test-bitcoincore-rpc/src/lib.rs]: bc8d25dd602aeee8665af3717c1fd6284975bbaf

File: tests/index.rs (hunk 20 above)
Comment (Muhammad4599) [reply tests/index.rs]: bc8d25dd602aeee8665af3717c1fd6284975bbaf

**Pull request merged**